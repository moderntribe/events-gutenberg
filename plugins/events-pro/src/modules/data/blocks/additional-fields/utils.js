export const FIELD_TYPES = {
	CHECKBOX: 'CHECKBOX',
	TEXT: 'TEXT',
	DROP_DOWN: 'DROP_DOWN',
	URL: 'URL',
	RADIO: 'RADIO',
	TEXT_AREA: 'TEXT_AREA',
};
