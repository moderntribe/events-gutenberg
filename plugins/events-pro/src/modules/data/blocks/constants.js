//
// ─── BLOCK TYPES ────────────────────────────────────────────────────────────────
//


export const RECURRING = 'recurring';
export const EXCEPTION = 'exception';

export const BLOCK_TYPES = [ RECURRING, EXCEPTION ];

//
// ─── STATE KEYS ─────────────────────────────────────────────────────────────────
//

export const KEY_TYPE = 'type';
export const KEY_ALL_DAY = 'all_day';
export const KEY_MULTI_DAY = 'multi_day';
export const KEY_START_TIME = 'start_time';
export const KEY_END_TIME = 'end_time';
export const KEY_START_DATE = 'start_date';
export const KEY_LIMIT = 'limit';
export const KEY_LIMIT_TYPE = 'limit_type';
export const KEY_BETWEEN = 'between';
export const KEY_TIMEZONE = 'timezone';
