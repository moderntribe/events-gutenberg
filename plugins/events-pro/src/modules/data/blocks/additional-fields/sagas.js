/**
 * External dependencies
 */
import { takeEvery, put, select, all } from 'redux-saga/effects';
import noop from 'lodash/noop';
import isEmpty from 'lodash/isEmpty';
import identity from 'lodash/identity';

/**
 * Internal dependencies
 */
import { actions, types, selectors } from '@moderntribe/events-pro/data/blocks/additional-fields';
import { FIELD_TYPES } from '@moderntribe/events-pro/blocks/additional-fields/utils';

export function* setInitialState( props ) {
	const { payload = {} } = props;
	const { name = '', get = noop } = payload;
	const type = get( 'type' );
	const value = get( 'value' );

	yield all( [
		put( actions.addField( name ) ),
		put( actions.setFieldIsPristine( name, get( 'isPristine' ) ) ),
		put( actions.setFieldType( name, type ) ),
		put( actions.setFieldLabel( name, get( 'label' ) ) ),
		put( actions.setFieldOptions( name, get( 'options' ) ) ),
		put( actions.setFieldValue( name, get( 'value' ) ) ),
	] );
}

export function* setPristineState( props ) {
	const { payload = {} } = props;
	const { name = '' } = payload;
	const value  = yield select( selectors.getFieldValue, payload );
	const isPristine = value === null || '' === value || isEmpty( value );
	yield put( actions.setFieldIsPristine( name, isPristine ) );
}

export function* appendFieldValue( props ) {
	const { payload = {} } = props;
	const { name = '', value } = payload;
	const current = yield select( selectors.getFieldCheckboxValue, payload );
	const newValue = [ ...current, value ]
		.filter( identity )
		.join( '|' );
	yield put( actions.setFieldValue( name, newValue ) );
}

export function* removeFieldValue( props ) {
	const { payload = {} } = props;
	const { name = '', value } = payload;
	const current = yield select( selectors.getFieldCheckboxValue, payload );
	const newValue = current
		.filter( ( text ) => text !== value )
		.join( '|' );
	yield put( actions.setFieldValue( name, newValue ) );
}


export default function* watchers() {
	yield takeEvery( types.SET_ADDITIONAL_FIELD_INITIAL_STATE, setInitialState );
	yield takeEvery( types.SET_ADDITIONAL_FIELD_CHANGE, setPristineState );
	yield takeEvery( types.APPEND_ADDITIONAL_FIELD_VALUE, appendFieldValue );
	yield takeEvery( types.REMOVE_ADDITIONAL_FIELD_VALUE, removeFieldValue );
}
