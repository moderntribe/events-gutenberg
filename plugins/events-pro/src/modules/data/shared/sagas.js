/**
 * External dependencies
 */
import { put, select, call, all } from 'redux-saga/effects';

/**
 * Internal dependencies
 */
import * as constants from '@moderntribe/events-pro/data/blocks/constants';
import * as recurringConstants from '@moderntribe/events-pro/data/blocks/recurring/constants';
import { moment as momentUtil, time } from '@moderntribe/common/utils';
import { selectors as datetime } from '@moderntribe/events/data/blocks/datetime';

const {
	KEY_TYPE,
	KEY_ALL_DAY,
	KEY_MULTI_DAY,
	KEY_START_TIME,
	KEY_END_TIME,
	KEY_START_DATE,
	KEY_START_DATE_INPUT,
	KEY_START_DATE_OBJ,
	KEY_END_DATE,
	KEY_END_DATE_INPUT,
	KEY_END_DATE_OBJ,
	KEY_LIMIT,
	KEY_LIMIT_DATE_INPUT,
	KEY_LIMIT_DATE_OBJ,
	KEY_LIMIT_TYPE,
	KEY_BETWEEN,
	KEY_DAYS,
	KEY_WEEK,
	KEY_DAY,
	KEY_MONTH,
	KEY_TIMEZONE,
	KEY_MULTI_DAY_SPAN,
} = constants;

const {
	toMoment,
	toDate,
	toDatabaseDate,
	toDatabaseTime,
} = momentUtil;

const {
	MINUTE_IN_SECONDS,
	HALF_HOUR_IN_SECONDS,
	HOUR_IN_SECONDS,
	DAY_IN_SECONDS,
	TIME_FORMAT_HH_MM,
	toSeconds,
	fromSeconds,
} = time;

export function* handleAddition( { actions } ) {
	const start = yield select( datetime.getStart );
	const end = yield select( datetime.getEnd );
	const allDay = yield select( datetime.getAllDay );
	const multiDay = yield select( datetime.getMultiDay );
	const timezone = yield select( datetime.getTimeZone );

	const startMoment = yield call( toMoment, start );
	const endMoment = yield call( toMoment, end );

	const startWeekday = yield call( [ startMoment, 'isoWeekday' ] );

	const startDate = yield call( toDatabaseDate, startMoment );
	const startTime = yield call( toDatabaseTime, startMoment );
	const endDate = yield call( toDatabaseDate, endMoment );
	const endTime = yield call( toDatabaseTime, endMoment );

	const startDateInput = yield call( toDate, startMoment );
	const startDateObj = new Date( startDateInput );
	const endDateInput = yield call( toDate, endMoment );
	const endDateObj = new Date( endDateInput );

	yield put( actions.add( {
		[ KEY_TYPE ]: recurringConstants.SINGLE,
		[ KEY_ALL_DAY ]: allDay,
		[ KEY_MULTI_DAY ]: multiDay,
		[ KEY_START_DATE ]: startDate,
		[ KEY_START_DATE_INPUT ]: startDateInput,
		[ KEY_START_DATE_OBJ ]: startDateObj,
		[ KEY_START_TIME ]: startTime,
		[ KEY_END_DATE ]: endDate,
		[ KEY_END_DATE_INPUT ]: endDateInput,
		[ KEY_END_DATE_OBJ ]: endDateObj,
		[ KEY_END_TIME ]: endTime,
		[ KEY_BETWEEN ]: 1,
		[ KEY_LIMIT_TYPE ]: recurringConstants.COUNT,
		[ KEY_LIMIT ]: 7,
		[ KEY_LIMIT_DATE_INPUT ]: endDateInput,
		[ KEY_LIMIT_DATE_OBJ ]: endDateObj,
		[ KEY_DAYS ]: [ startWeekday ],
		[ KEY_WEEK ]: recurringConstants.FIRST,
		[ KEY_DAY ]: startWeekday,
		[ KEY_MONTH ]: [],
		[ KEY_TIMEZONE ]: timezone,
		[ KEY_MULTI_DAY_SPAN ]: recurringConstants.NEXT_DAY,
	} ) );
}

export function* handleTimeChange( { actions, selectors }, action, key ) {
	const payloadTime = action.payload[ key ];
	const isAllDay = payloadTime === 'all-day';
	const isMultiDay = yield select( selectors.getMultiDay, action );

	if ( isAllDay ) {
		yield put(
			actions.sync( action.index, {
				[ KEY_ALL_DAY ]: isAllDay,
				[ KEY_START_TIME ]: '00:00:00',
				[ KEY_END_TIME ]: '23:59:59',
			} )
		);
	} else if ( ! isMultiDay ) {
		const isStartTime = key === KEY_START_TIME;
		const isEndTime = key === KEY_END_TIME;

		const startTime = isStartTime
			? payloadTime
			: yield select( selectors.getStartTimeNoSeconds, action );

		const endTime = isEndTime
			? payloadTime
			: yield select( selectors.getEndTimeNoSeconds, action );

		// This put needs to be here to prevent incorrect time formats when 'adjusting'
		// as it will not adjust in all cases
		yield put(
			actions.sync( action.index, {
				[ KEY_ALL_DAY ]: isAllDay,
				[ key ]: `${ payloadTime }:00`,
			} )
		);

		isStartTime
			? yield call( preventEndTimeBeforeStartTime, { actions }, { startTime, endTime }, action )
			: yield call( preventStartTimeAfterEndTime, { actions }, { startTime, endTime }, action );
	} else {
		yield put(
			actions.sync( action.index, {
				[ KEY_ALL_DAY ]: isAllDay,
				[ key ]: `${ payloadTime }:00`,
			} )
		);
	}
}

export function* handleMultiDayChange( { actions, selectors }, action, key ) {
	const isMultiDay = action.payload[ key ];

	if ( ! isMultiDay ) {
		const startTime = yield select( selectors.getStartTimeNoSeconds, action );
		const endTime = yield select( selectors.getEndTimeNoSeconds, action );

		yield call( preventEndTimeBeforeStartTime, { actions }, { startTime, endTime }, action );
	}
}

/**
 * Prevents end time from being before start time.
 * Should only prevent when not a multi-day event.
 *
 * @export
 * @param {Object} { actions } Actions for syncing
 * @param {Object} { startTime, endTime } Start and end time
 * @param {Object} action Action received
 */
export function* preventEndTimeBeforeStartTime( { actions }, { startTime, endTime }, action ) {
	let startTimeSeconds = yield call( toSeconds, startTime, TIME_FORMAT_HH_MM );
	let endTimeSeconds = yield call( toSeconds, endTime, TIME_FORMAT_HH_MM );

	// If end time is earlier than start time, fix time
	if ( endTimeSeconds <= startTimeSeconds ) {
		// If there is less than half an hour left in the day, roll back one hour
		if ( startTimeSeconds + HALF_HOUR_IN_SECONDS >= DAY_IN_SECONDS ) {
			startTimeSeconds -= HOUR_IN_SECONDS;
		}

		endTimeSeconds = startTimeSeconds + HALF_HOUR_IN_SECONDS;

		const adjustedStartTime = yield call( fromSeconds, startTimeSeconds, TIME_FORMAT_HH_MM );
		const adjustedEndTime = yield call( fromSeconds, endTimeSeconds, TIME_FORMAT_HH_MM );

		yield put(
			actions.sync( action.index, {
				[ KEY_START_TIME ]: `${ adjustedStartTime }:00`,
				[ KEY_END_TIME ]: `${ adjustedEndTime }:00`,
			} )
		);
	}
}

/**
 * Prevents start time from appearing ahead of end time.
 * Should only prevent when not a multi-day event.
 *
 * @export
 * @param {Object} { actions } Actions for syncing
 * @param {Object} { startTime, endTime } Start and end time
 * @param {Object} action Action received
 */
export function* preventStartTimeAfterEndTime( { actions }, { startTime, endTime }, action ) {
	let startTimeSeconds = yield call( toSeconds, startTime, TIME_FORMAT_HH_MM );
	let endTimeSeconds = yield call( toSeconds, endTime, TIME_FORMAT_HH_MM );

	if ( startTimeSeconds >= endTimeSeconds ) {
		startTimeSeconds = Math.max( endTimeSeconds - HALF_HOUR_IN_SECONDS, 0 );
		endTimeSeconds = Math.max( startTimeSeconds + MINUTE_IN_SECONDS, endTimeSeconds );

		const adjustedStartTime = yield call( fromSeconds, startTimeSeconds, TIME_FORMAT_HH_MM );
		const adjustedEndTime = yield call( fromSeconds, endTimeSeconds, TIME_FORMAT_HH_MM );

		yield put(
			actions.sync( action.index, {
				[ KEY_START_TIME ]: `${ adjustedStartTime }:00`,
				[ KEY_END_TIME ]: `${ adjustedEndTime }:00`,
			} )
		);
	}
}

export function* handleWeekChange( { actions, selectors }, action, key ) {
	const payloadWeek = action.payload[ key ];
	const weekWasNull = ! ( yield select( selectors.getWeek, action ) );

	if ( payloadWeek && weekWasNull ) {
		yield put(
			actions.sync( action.index, {
				[ key ]: payloadWeek,
				[ KEY_DAY ]: 1,
			} )
		);
	}
}

export function* handleLimitTypeChange( { actions }, action, key ) {
	const value = action.payload[ key ];
	const isDate = value === recurringConstants.DATE;
	const isCount = value === recurringConstants.COUNT;

	if ( isDate ) {
		const start = yield select( datetime.getStart );
		const startMoment = yield call( toMoment, start );
		const startDate = yield call( toDatabaseDate, startMoment );
		yield put(
			actions.sync( action.index, {
				[ constants.KEY_LIMIT ]: startDate,
			} )
		);
	} else if ( isCount ) {
		yield put(
			actions.sync( action.index, {
				[ constants.KEY_LIMIT ]: 1,
			} )
		);
	} else {
		// Never ending
		yield put(
			actions.sync( action.index, {
				[ constants.KEY_LIMIT ]: null,
			} )
		);
	}
}

export function* handleTimezoneChange( { actions }, action, key ) {
	yield put(
		actions.sync( action.index, {
			[ constants.KEY_TIMEZONE ]: action.payload[ key ],
		} )
	);
}
