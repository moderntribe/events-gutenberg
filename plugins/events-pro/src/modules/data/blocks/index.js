/**
 * Internal dependencies
 */
import * as constants from './constants';
import reducer from './reducer';
import './sagas';

export default reducer;
export { constants };
