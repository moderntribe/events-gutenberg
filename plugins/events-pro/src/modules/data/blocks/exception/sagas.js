/**
 * External dependencies
 */
import { takeEvery, put, select } from 'redux-saga/effects';

/**
 * Internal dependencies
 */
import * as exception from '@moderntribe/events-pro/data/blocks/exception';
import * as recurring from '@moderntribe/events-pro/data/blocks/recurring';
import * as ui from '@moderntribe/events-pro/data/ui';
import { moment as momentUtil } from '@moderntribe/common/utils';
import { selectors as datetime } from '@moderntribe/events/data/blocks/datetime';

export function* handleExceptionRemoval() {
	const exceptions = yield select( exception.selectors.getExceptions );

	if ( ! exceptions.length ) {
		yield put( ui.actions.hideExceptionPanel() );
	}
}

export function* handleExceptionAddition() {
	const start = yield select( datetime.getStart );
	const end = yield select( datetime.getEnd );
	const allDay = yield select( datetime.getAllDay );
	const multiDay = yield select( datetime.getMultiDay );
	const timezone = yield select( datetime.getTimeZone );

	const startMoment = momentUtil.toMoment( start );
	const endMoment = momentUtil.toMoment( end );

	const startDate = momentUtil.toDate( startMoment );
	const startTime = momentUtil.toTime24Hr( startMoment );
	const endDate = momentUtil.toDate( endMoment );
	const endTime = momentUtil.toTime24Hr( endMoment );

	yield put( exception.actions.addException( {
		type: recurring.constants.SINGLE,
		all_day: allDay,
		multi_day: multiDay,
		start_date: startDate,
		start_time: startTime,
		end_date: endDate,
		end_time: endTime,
		between: 1,
		limit_type: recurring.constants.COUNT,
		limit: 7,
		days: [],
		week: recurring.constants.FIRST,
		day: 1,
		month: [],
		timezone,
	} ) );
}

export default function* watchers() {
	yield takeEvery( [ exception.types.REMOVE_EXCEPTION ], handleExceptionRemoval );
	yield takeEvery( [ exception.types.ADD_EXCEPTION_FIELD ], handleExceptionAddition );
}
