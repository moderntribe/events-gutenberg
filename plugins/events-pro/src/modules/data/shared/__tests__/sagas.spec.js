/**
 * External dependencies
 */
import { put, select, call } from 'redux-saga/effects';

/**
 * Internal Dependencies
 */
import * as constants from '@moderntribe/events-pro/data/blocks/constants';
import * as recurringConstants from '@moderntribe/events-pro/data/blocks/recurring/constants';
import { moment as momentUtil, time } from '@moderntribe/common/utils';
import { selectors as datetime } from '@moderntribe/events/data/blocks/datetime';
import * as sagas from '../sagas';

describe( 'Shared recurrence sagas', () => {
	let actions, selectors, args;

	beforeEach( () => {
		actions = {
			sync: jest.fn( obj => obj ),
			add: jest.fn( obj => obj ),
		};
		selectors = {
			getStartTimeNoSeconds: jest.fn( () => '12:00' ),
			getEndTimeNoSeconds: jest.fn( () => '15:00' ),
			getWeek: jest.fn(),
			getMultiDay: jest.fn(),
		};
		args = {
			start_date: '2018-01-01 12:00:00',
			end_date: '2018-01-02 12:00:00',
			all_day: false,
			multi_day: false,
			timezone: 'UTC+0',
		};
	} );

	describe( 'handleAddition', () => {
		const _tempDate = global.Date;
		beforeAll( () => {
			global.Date = class _Date {
				constructor( date ) {
					return date;
				}
			};
		} );

		afterAll( () => {
			global.Date = _tempDate;
		} );
		it( 'should add rule', () => {
			const gen = sagas.handleAddition( { actions } );
			expect( gen.next().value ).toEqual(
				select( datetime.getStart )
			);
			expect( gen.next( args.start_date ).value ).toEqual(
				select( datetime.getEnd )
			);
			expect( gen.next( args.end_date ).value ).toEqual(
				select( datetime.getAllDay )
			);
			expect( gen.next( args.all_day ).value ).toEqual(
				select( datetime.getMultiDay )
			);
			expect( gen.next( args.multi_day ).value ).toEqual(
				select( datetime.getTimeZone )
			);

			expect( gen.next( args.timezone ).value ).toEqual(
				call( momentUtil.toMoment, args.start_date )
			);
			expect( gen.next( args.start_date ).value ).toEqual(
				call( momentUtil.toMoment, args.end_date )
			);

			expect( gen.next( args.end_date ).value ).toEqual(
				call( momentUtil.toDatabaseDate, args.start_date )
			);
			expect( gen.next( args.start_date ).value ).toEqual(
				call( momentUtil.toDatabaseTime, args.start_date )
			);

			expect( gen.next( args.start_date ).value ).toEqual(
				call( momentUtil.toDatabaseDate, args.end_date )
			);
			expect( gen.next( args.end_date ).value ).toEqual(
				call( momentUtil.toDatabaseTime, args.end_date )
			);

			expect( gen.next( args.end_date ).value ).toEqual(
				call( momentUtil.toDate, args.start_date )
			);
			expect( gen.next( args.start_date ).value ).toEqual(
				call( momentUtil.toDate, args.end_date )
			);

			expect( gen.next().value ).toEqual(
				put(
					actions.add( {
						[ constants.KEY_TYPE ]: recurringConstants.SINGLE,
						[ constants.KEY_ALL_DAY ]: args.all_day,
						[ constants.KEY_MULTI_DAY ]: args.multi_day,
						[ constants.KEY_START_DATE ]: args.start_date,
						[ constants.KEY_START_DATE_INPUT ]: args.start_date,
						[ constants.KEY_START_DATE_OBJ ]: new Date( '2018-01-01T12:00:00.000Z' ),
						[ constants.KEY_START_TIME ]: args.start_date,
						[ constants.KEY_END_DATE ]: args.end_date,
						[ constants.KEY_END_DATE_INPUT ]: undefined,
						[ constants.KEY_END_DATE_OBJ ]: new Date( NaN ),
						[ constants.KEY_END_TIME ]: args.end_date,
						[ constants.KEY_BETWEEN ]: 1,
						[ constants.KEY_LIMIT_TYPE ]: recurringConstants.COUNT,
						[ constants.KEY_LIMIT ]: 7,
						[ constants.KEY_LIMIT_DATE_INPUT ]: undefined,
						[ constants.KEY_LIMIT_DATE_OBJ ]: new Date( NaN ),
						[ constants.KEY_DAYS ]: [],
						[ constants.KEY_WEEK ]: recurringConstants.FIRST,
						[ constants.KEY_DAY ]: 1,
						[ constants.KEY_MONTH ]: [],
						[ constants.KEY_TIMEZONE ]: args.timezone,
						[ constants.KEY_MULTI_DAY_SPAN ]: recurringConstants.NEXT_DAY,
					} )
				)
			);
		} );
	} );

	describe( 'handleTimeChange', () => {
		it( 'should handle time change when all-day', () => {
			const action = { payload: {
				[ constants.KEY_END_TIME ]: 'all-day',
			}, index: 0 };
			const gen = sagas.handleTimeChange( { actions, selectors }, action, constants.KEY_END_TIME );

			expect( gen.next().value ).toEqual(
				select( selectors.getMultiDay, action )
			);

			expect( gen.next().value ).toEqual(
				put( actions.sync( 0, {
					[ constants.KEY_ALL_DAY ]: true,
					[ constants.KEY_START_TIME ]: '00:00:00',
					[ constants.KEY_END_TIME ]: '23:59:59',
				} ) )
			);

			expect( gen.next().done ).toEqual( true );
		} );
		describe( 'handle non-multi', () => {
			it( 'should handle start time changes', () => {
				const action = { payload: {
					[ constants.KEY_START_TIME ]: '12:00',
				}, index: 0 };
				const gen = sagas.handleTimeChange( { actions, selectors }, action, constants.KEY_START_TIME );

				expect( gen.next().value ).toEqual(
					select( selectors.getMultiDay, action )
				);

				expect( gen.next().value ).toEqual(
					select( selectors.getEndTimeNoSeconds, action )
				);

				expect( gen.next( '15:00' ).value ).toEqual(
					put( actions.sync( 0, {
						[ constants.KEY_ALL_DAY ]: false,
						[ constants.KEY_START_TIME ]: '12:00:00',
					} ) )
				);

				expect( gen.next().value ).toEqual(
					call( sagas.preventEndTimeBeforeStartTime, { actions }, {
						startTime: action.payload[ constants.KEY_START_TIME ],
						endTime: '15:00',
					}, action )
				);

				expect( gen.next().done ).toEqual( true );
			} );
			it( 'should handle end time changes', () => {
				const action = { payload: {
					[ constants.KEY_END_TIME ]: '12:00',
				}, index: 0 };
				const gen = sagas.handleTimeChange( { actions, selectors }, action, constants.KEY_END_TIME );

				expect( gen.next().value ).toEqual(
					select( selectors.getMultiDay, action )
				);

				expect( gen.next().value ).toEqual(
					select( selectors.getStartTimeNoSeconds, action )
				);

				expect( gen.next( '15:00' ).value ).toEqual(
					put( actions.sync( 0, {
						[ constants.KEY_ALL_DAY ]: false,
						[ constants.KEY_END_TIME ]: '12:00:00',
					} ) )
				);

				expect( gen.next().value ).toEqual(
					call( sagas.preventStartTimeAfterEndTime, { actions }, {
						startTime: '15:00',
						endTime: action.payload[ constants.KEY_END_TIME ],
					}, action )
				);

				expect( gen.next().done ).toEqual( true );
			} );
		} );
		it( 'should handle multi day changes', () => {
			const action = { payload: {
				[ constants.KEY_END_TIME ]: '12:00',
			}, index: 0 };
			const gen = sagas.handleTimeChange( { actions, selectors }, action, constants.KEY_END_TIME );

			expect( gen.next().value ).toEqual(
				select( selectors.getMultiDay, action )
			);

			expect( gen.next( true ).value ).toEqual(
				put( actions.sync( 0, {
					[ constants.KEY_ALL_DAY ]: false,
					[ constants.KEY_END_TIME ]: '12:00:00',
				} ) )
			);

			expect( gen.next().done ).toEqual( true );
		} );
	} );

	describe( 'handleMultiDayChange', () => {
		it( 'should do nothing', () => {
			const gen = sagas.handleMultiDayChange( { actions, selectors }, {
				payload: {
					[ constants.KEY_MULTI_DAY ]: true,
				},
				index: 0,
			}, constants.KEY_MULTI_DAY );

			gen.next().value;

			expect( gen.next().done ).toEqual( true );
		} );
		it( 'should handle when not multi day', () => {
			const startTime = '12:00';
			const endTime = '12:00';
			const action = {
				payload: {
					[ constants.KEY_MULTI_DAY ]: false,
				},
				index: 0,
			};

			const gen = sagas.handleMultiDayChange( { actions, selectors }, action, constants.KEY_MULTI_DAY );

			expect( gen.next().value ).toEqual(
				select( selectors.getStartTimeNoSeconds, action )
			);
			expect( gen.next( startTime ).value ).toEqual(
				select( selectors.getEndTimeNoSeconds, action )
			);
			expect( gen.next( endTime ).value ).toEqual(
				call( sagas.preventEndTimeBeforeStartTime, { actions }, { startTime, endTime }, action )
			);
			expect( gen.next().done ).toEqual( true );
		} );
	} );

	describe( 'handleWeekChange', () => {
		it( 'should sync week', () => {
			const action = {
				index: 0,
				payload: {
					[ constants.KEY_WEEK ]: 1,
				},
			};
			const gen = sagas.handleWeekChange( { actions, selectors }, action, constants.KEY_WEEK );

			expect( gen.next().value ).toEqual(
				select( selectors.getWeek, action )
			);

			expect( gen.next( false ).value ).toEqual(
				put(
					actions.sync( action.index, {
						[ constants.KEY_WEEK ]: 1,
						[ constants.KEY_DAY ]: 1,
					} )
				)
			);

			expect( gen.next().done ).toBeTruthy();
		} );
	} );

	describe( 'preventEndTimeBeforeStartTime', () => {
		it( 'should adjust time when start and end time are the same', () => {
			const startTime = '12:00';
			const endTime = '12:00';
			const action = {
				payload: {
					[ constants.KEY_MULTI_DAY ]: false,
				},
				index: 0,
			};

			const gen = sagas.preventEndTimeBeforeStartTime( { actions }, { startTime, endTime }, action );

			expect( gen.next( endTime ).value ).toEqual(
				call( time.toSeconds, startTime, time.TIME_FORMAT_HH_MM )
			);
			expect( gen.next( 43200 ).value ).toEqual(
				call( time.toSeconds, endTime, time.TIME_FORMAT_HH_MM )
			);
			expect( gen.next( 43200 ).value ).toEqual(
				call( time.fromSeconds, 43200, time.TIME_FORMAT_HH_MM )
			);
			expect( gen.next( startTime ).value ).toEqual(
				call( time.fromSeconds, 45000, time.TIME_FORMAT_HH_MM )
			);
			expect( gen.next( '13:00' ).value ).toEqual(
				put( actions.sync( 0, {
					[ constants.KEY_START_TIME ]: `12:00:00`,
					[ constants.KEY_END_TIME ]: `13:00:00`,
				} ) )
			);
		} );
		it( 'should roll back start time when exceeds day time', () => {
			const startTime = '23:59';
			const endTime = '23:59';
			const action = {
				payload: {
					[ constants.KEY_MULTI_DAY ]: false,
				},
				index: 0,
			};

			const gen = sagas.preventEndTimeBeforeStartTime( { actions }, { startTime, endTime }, action );

			expect( gen.next( endTime ).value ).toEqual(
				call( time.toSeconds, startTime, time.TIME_FORMAT_HH_MM )
			);
			expect( gen.next( 86340 ).value ).toEqual(
				call( time.toSeconds, endTime, time.TIME_FORMAT_HH_MM )
			);
			expect( gen.next( 86340 ).value ).toEqual(
				call( time.fromSeconds, 82740, time.TIME_FORMAT_HH_MM )
			);
			expect( gen.next( startTime ).value ).toEqual(
				call( time.fromSeconds, 84540, time.TIME_FORMAT_HH_MM )
			);
			expect( gen.next().value ).toEqual(
				put( actions.sync( 0, {
					[ constants.KEY_START_TIME ]: `12:00:00`,
					[ constants.KEY_END_TIME ]: `13:00:00`,
				} ) )
			);
		} );
	} );
	describe( 'preventStartTimeAfterEndTime', () => {
		it( 'should adjust time when start and end time are the same', () => {
			const startTime = '12:00';
			const endTime = '12:00';
			const action = {
				payload: {
					[ constants.KEY_MULTI_DAY ]: false,
				},
				index: 0,
			};

			const gen = sagas.preventStartTimeAfterEndTime( { actions }, { startTime, endTime }, action );

			expect( gen.next().value ).toEqual(
				call( time.toSeconds, startTime, time.TIME_FORMAT_HH_MM )
			);
			expect( gen.next( 43200 ).value ).toEqual(
				call( time.toSeconds, endTime, time.TIME_FORMAT_HH_MM )
			);


			expect( gen.next( 43200 ).value ).toEqual(
				call( time.fromSeconds, 41400, time.TIME_FORMAT_HH_MM )
			);
			expect( gen.next( startTime ).value ).toEqual(
				call( time.fromSeconds, 43200, time.TIME_FORMAT_HH_MM )
			);


			expect( gen.next( '11:30' ).value ).toEqual(
				put( actions.sync( 0, {
					[ constants.KEY_START_TIME ]: `11:30:00`,
					[ constants.KEY_END_TIME ]: `12:00:00`,
				} ) )
			);
		} );
		it( 'should handle when start time would be on previous day', () => {
			const startTime = '00:00';
			const endTime = '00:00';
			const action = {
				payload: {
					[ constants.KEY_MULTI_DAY ]: false,
				},
				index: 0,
			};

			const gen = sagas.preventStartTimeAfterEndTime( { actions }, { startTime, endTime }, action );

			expect( gen.next( endTime ).value ).toEqual(
				call( time.toSeconds, startTime, time.TIME_FORMAT_HH_MM )
			);
			expect( gen.next( 0 ).value ).toEqual(
				call( time.toSeconds, endTime, time.TIME_FORMAT_HH_MM )
			);
			expect( gen.next( 0 ).value ).toEqual(
				call( time.fromSeconds, 0, time.TIME_FORMAT_HH_MM )
			);
			expect( gen.next( startTime ).value ).toEqual(
				call( time.fromSeconds, 60, time.TIME_FORMAT_HH_MM )
			);
			expect( gen.next().value ).toEqual(
				put( actions.sync( 0, {
					[ constants.KEY_START_TIME ]: `00:00:00`,
					[ constants.KEY_END_TIME ]: `00:01:00`,
				} ) )
			);
		} );
	} );

	describe( 'handleLimitTypeChange', () => {
		let action, payload;
		beforeEach( () => {
			payload = {};
			action = {
				index: 0,
				payload,
			};
		} );

		it( 'should handle date changes', () => {
			payload[ constants.KEY_LIMIT ] = recurringConstants.DATE;
			const gen = sagas.handleLimitTypeChange( { actions }, action, constants.KEY_LIMIT );

			expect( gen.next().value ).toEqual(
				select( datetime.getStart )
			);

			expect( gen.next( '2018-01-01' ).value ).toEqual(
				call( momentUtil.toMoment, '2018-01-01' )
			);

			expect( gen.next( '2018-01-01' ).value ).toEqual(
				call( momentUtil.toDatabaseDate, '2018-01-01' )
			);

			expect( gen.next().value ).toEqual(
				put( actions.sync( action.index, {
					[ constants.KEY_LIMIT ]: '2018-01-01',
				} ) )
			);

			expect( gen.next().done ).toBeTruthy();
		} );

		it( 'should handle count changes', () => {
			payload[ constants.KEY_LIMIT ] = recurringConstants.COUNT;
			const gen = sagas.handleLimitTypeChange( { actions }, action, constants.KEY_LIMIT );

			expect( gen.next().value ).toEqual(
				put( actions.sync( action.index, {
					[ constants.KEY_LIMIT ]: 1,
				} ) )
			);
			expect( gen.next().done ).toBeTruthy();
		} );

		it( 'should handle never ending changes', () => {
			const gen = sagas.handleLimitTypeChange( { actions }, action, constants.KEY_LIMIT );

			expect( gen.next().value ).toEqual(
				put( actions.sync( action.index, {
					[ constants.KEY_LIMIT ]: null,
				} ) )
			);

			expect( gen.next().done ).toBeTruthy();
		} );
	} );

	describe( 'handleTimezoneChange', () => {
		it( 'should sync timezone', () => {
			const action = {
				index: 0,
				payload: {
					[ constants.KEY_TIMEZONE ]: 'UTC+0',
				},
			};
			const gen = sagas.handleTimezoneChange( { actions }, action, constants.KEY_TIMEZONE );

			expect( gen.next().value ).toEqual(
				put( actions.sync( action.index, {
					[ constants.KEY_TIMEZONE ]: action.payload[ constants.KEY_TIMEZONE ],
				} ) )
			);

			expect( gen.next().done ).toBeTruthy();
		} );
	} );
} );
