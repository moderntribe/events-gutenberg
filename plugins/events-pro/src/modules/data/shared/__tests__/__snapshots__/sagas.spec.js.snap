// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shared recurrence sagas handleAddition should add rule 1`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "_end_date_input": undefined,
      "_end_date_obj": Date { NaN },
      "_limit_date_input": undefined,
      "_limit_date_obj": Date { NaN },
      "_start_date_input": "2018-01-01 12:00:00",
      "_start_date_obj": 2018-01-01T12:00:00.000Z,
      "all_day": false,
      "between": 1,
      "day": 1,
      "days": Array [
        1,
      ],
      "end_date": "2018-01-02 12:00:00",
      "end_time": "2018-01-02 12:00:00",
      "limit": 7,
      "limit_type": "count",
      "month": Array [],
      "multi_day": false,
      "multi_day_span": "next_day",
      "start_date": "2018-01-01 12:00:00",
      "start_time": "2018-01-01 12:00:00",
      "timezone": "UTC+0",
      "type": "single",
      "week": "first",
    },
    "channel": null,
  },
}
`;
